modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1976,27 +2084,71 @@ def _execute_arrow(self):
         pyarrow.Table
             The resulting table.
         """
-        if isinstance(self._op, FrameNode):
-            if self._partitions.size == 0:
-                return pyarrow.Table.from_pandas(pd.DataFrame({}))
+        result = None
+        stack = [self]
+
+        while stack:
+            frame = stack.pop()
+
+            if callable(frame):
+                frame()
+            elif isinstance(frame._op, FrameNode):
+                if frame._partitions.size == 0:
+                    result = pyarrow.Table.from_pandas(pd.DataFrame({}))
+                else:
+                    assert frame._partitions.size == 1
+                    result = frame._partitions[0][0].get()
+            elif isinstance(frame._op, MaskNode):
+
+                def slice(positions=frame._op.row_positions):
+                    nonlocal result
+                    result = self._arrow_row_slice(result, positions)
+
+                stack.append(slice)
+                stack.append(frame._op.input[0])
+            elif isinstance(frame._op, TransformNode):
+
+                def select(exprs=frame._op.exprs):
+                    nonlocal result
+                    result = self._arrow_select(result, exprs)
+
+                stack.append(select)
+                stack.append(frame._op.input[0])
+            elif isinstance(frame._op, UnionNode):
+
+                def union(op=frame._op, tables={}, input=iter(frame._op.input)):
+                    nonlocal result
+
+                    if (i := next(input, None)) is None:
+                        result = self._union_all_arrow(
+                            op.input, op.join, op.sort, op.ignore_index, tables
+                        )
+                    else:
+
+                        def add_result(f=i):
+                            tables[f] = result
+
+                        stack.append(frame if callable(frame) else union)
+                        stack.append(add_result)
+                        stack.append(i)
```

Comment:
doesn't this only add the first table to the stack? [`UnionNode.input` is a list](https://github.com/modin-project/modin/blob/8817093cb2f16be3709513b1517d8a4bc679900c/modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py#L727-L728), so what happens to the other tables from the `_op.input`?

Reply:
When the function union() is called, it adds self, add_result and the next frame to the stack. This is repeated until we reach the end of the iterator.